### DISCUSSION TITLE
Expand explicit implementation to be more flexible

### DISCUSSION THREAD

**Discussion by HaloFour** [dotnet/csharplang#7210] [Language Ideas]
Currently C# offers a somewhat limited form of explicit implementation syntax that allows the implementing class to "hide" the implemented member.  If you have a use case where you want to hide one member in order to favor another member with a different name, you are required to use explicit implementation as a bridge method to call that other member:

```cs
public class Foo : IDisposable
{
    void IDisposable.Dispose() => this.Close();

    protected void Close() { ... }
}
```

The runtime has actually always offered a much more flexible way to express this in IL, where an implementing member can have any accessibility or name as long as the rest of the signature was identical.  VB.NET does expose this feature:

```vb
Public Class Foo
    Implements IDisposable

    ' The Close method explicitly satisfies the IDisposable contract
    Protected Sub Close() Implements IDisposable.Dispose
    End Sub
End Class
```

It could be useful in these scenarios if C# offered the same functionality, perhaps using the following syntax:

```cs
public class Foo : IDisposable
{
    protected void Close() : IDisposable.Dispose { ... }
}
```

**Comment by HaloFour** [comment]
I'd be curious how often explicit implementation is used specifically for these use cases, and if a feature like this would have been directly applicable instead.  The scenario in which this arises is probably somewhat rare, but that could also be because it's not currently possible in the language without that manual bridging via explicit implementation.

**Comment by FaustVX** [comment]
I like this idea, even if it's a niche one, but the suggested syntax looks weird to me :
```C#
protected void Close() : IDisposable.Dispose { ... }
```
It would look more C#-ish to me if we just allow accessibility modifier on explicit implementation
```C#
protected void IDisposable.Dispose() { ... }
```

**Reply by HaloFour** [reply]
> It would look more C#-ish to me if we just allow accessibility modifier on explicit implementation

How would you name the method `Close` in that case?

**Reply by FaustVX** [reply]
Ok, I didn't see that you have renamed the method.

**Reply by HaloFour** [reply]
I'm not married to the syntax.  It's largely adapted from the VB.NET syntax, switching `Implements` for `:`.  It feels somewhat C#ish to me in that it's similar to class inheritance.

**Comment by tfenise** [comment]
It could also save the overhead of one extra (virtual) call, compared to manual bridging.

**Comment by pinkfloydx33** [comment]
Starting out as a VB developer, I've definitely had several cases over the last decade where I've missed the ability to do this in C#. FWIW I can't recall the exaxt scenarios, just that they existed and I almost certainly used explicit implementation as a bridge. 

Would this be a nice to have? Absolutely. Would I want the team to prioritize this over extensions/roles, DUs and other long-term asks? Nope. Of course if it fell out of something else I wouldn't complain.

**Comment by yaakov-h** [comment]
Would you be able to fulfil multiple interfaces this way?

e.g.

```csharp
public IEnumerator<object> GetEnumerator() : IEnumerable.GetEnumerator, IEnumerable<object>.GetEnumerator
{
    throw null;
}
```

**Reply by HaloFour** [reply]
It seems that IL does permit multiple `.override` clauses for a method, so why not?

**Reply by pinkfloydx33** [reply]
Unless I'm misremembering (or doing something _horribly_ wrong), I'm pretty sure in VB at least that you'd still need two methods to implement `IEnumerable{T}.GetEnumerator` since the return types don't match. Whether this is a language restriction or a _CLR_ one I don't know. If the CLR supported it, it would certainly be nice to have. 

On the other hand, you _are_ able to implement `IEnumerator.Current` and `IEnumerable<T>.Current` with a single (explicit/renamed) member since they both return `object`. 

```vb
Public Class Foo 
    Implements IEnumerator(Of Object) 
    Public Readonly Property MySillyCurrent() As Object Implements IEnumerator(Of Object).Current, IEnumerator.Current 
        Get
           Return Nothing 
        End Get
    End Property
    Public Function MoveNext() As Boolean Implements IEnumerator.MoveNext
        Return False
    End Function
    Public Sub Reset Implements IEnumerator.Reset
    End Sub
    Public Sub Dispose Implements IDisposable.Dispose
    End Sub
End Class 
```

[SharpLab](https://sharplab.io/#v2:DYLgbgRgNALiBOBXAdlAJiA1AHwJIFsAHAe3hgGcACAZQE9yYBTfAWACgCSyq6HmA6AMLFgwRgGMYAS2LJy7TqQo16TfEJFjJMufwDijZI3hTx7dgAVEEYKcoBZYmkRjKF+MQDm8AIas2lIFu1rbiNNYOPlLIABQAlOxBSZTCciKM/ADqJkwAMtGMMQBEgDwbgCD7RQkBQQCiyGjhEOx1DY7OYuZsVjZ2gsA+5FQAYsTElIlBnGL4hsq4dYgzvjCkMQDyAGaUaxAAVhIwcePVgd2hlABKjD5ossC0bh6ExjAP9rTUUqK0gojw8LN4pQAIJUHb7SSUKbMWZUebIRbGHwreDrLbgg5xIR/AHIGBQKELJbI0jY/6zY7JIIGGATKlBK4wP7ISgAOWIMAAFtFPJT6S1KDS6ZQBe5iM8yLRhWc7EMUNpZA5iGBGKzGAAPGBA0GUABCozEPhZ0JmeLhRKRKP4jhVas1wqSjOZlCGPmA5EYwoFcuQCuQ0pCdmoESuHpgUKI01hhIRxKtocYtJOIvqjQDPTCwYglAAIlJyCQPRHCFGzVC8wXiOQfDYMhXC57kwKs81U30BlQgA===) 

But you still cannot implement `IEnumerable<string>.Current` and `IEnumerable.Current` with the same property... again, I believe since the return type doesn't match and I'm not sure if this is a CLR or language restriction. Apologies if I'm doing something wrong and just mucking up the syntax. And I realize we're talking about C# and _not_ VB... but I figured there might be good IL/CLR related reason it's not supported 

At the very least we should want parity...but supporting more cases would certainly be nice. 

BTW, I think your example is a bit confusing as it _seems_ (?) to be mixing `IEnumerable.GetEnumerator` with `IEnumerator.Current` (apologies if it's intentional and/or I'm missing the point)

**Reply by HaloFour** [reply]
Right, it couldn't work for `IEnumerable.GetEnumerator` and `IEnumerable<object>.GetEnumerator`, but it could work for `IEnumerator.Current` and `IEnumerator<object>.Current`.  Although I would be curious how covariant returns in the runtime work with explicit `.override` clauses like this....

VB.NET does allow implementing both interface members with a single method, if this feature were to be considered for C# I don't see why it wouldn't support the same, although admittedly it's probably a lot less useful.  Probably comes up more often when combining multiple non-generic interfaces that have some common members and you want them to have the same implementation.

**Comment by kyoyama-kazusa** [comment]
Hi.

How does this solve the overloading problem, e.g. an interface type contains at least two methods whose parameter lists are the only differ?

```csharp
interface I
{
    void Method();
    void Method(int param);
}
```

```csharp
public void Method() : I.Method // What method does 'I.Method' mean?
    => Console.WriteLine(42);
```

**Reply by yaakov-h** [reply]
I would think that the parameter list would still have to match. There's no way that `void Method()` would be compatible with `void Method(int)`.

**Reply by HaloFour** [reply]
Right, the signature of the overloading/implementing method must match.  So you could have the following:

```cs
public class C : I
{
    public void M1() : I.Method;
    public void M2(int param) : I.Method;
}
```

That's how it works today in VB.NET:

```vb
Interface I
    Sub Method()
    Sub Method(param As Integer)
End Interface

Class C
    Implements I
    
    Public Sub M1() Implements I.Method
    End Sub
    
    Public Sub M2(param As Integer) Implements I.Method
    End Sub
End Class
```
